In [1]:
import duckdb
import polars as pl
from polars import col as c
import polars.selectors as cs
from config import BASE_PARQUET_PATH, DATABASE_URL


## Hospital Analysis

Using the hospital data stored in the `hospital_directory` table of the duckdb database, perform the following analyses:

1. Determine the total number of hospitals processed `processed`
2. Determine the number of hospitals with files compliant with CMS requirements `compliant`
3. Determine the number of hospitals with drugs listed in the file `has_drugs`
4. Determine the number of hospitals with drugs and corresponding prices `has_prices`

In [2]:
def run_query(query: str) -> pl.LazyFrame:
    with duckdb.connect(DATABASE_URL) as con:
        return pl.DataFrame(con.execute(query=query).fetch_arrow_table()).lazy()

def base_query():
    sql = f"""select * from hospital_directory"""
    return run_query(sql)


In [3]:
(
base_query()
.select(cs.boolean().exclude('converted').sum())
.collect(engine='streaming')
.to_pandas()
)

processed  compliant  has_drugs  has_pricing
0       2992       2617       2122          998

In [4]:
(
base_query()
.select(cs.boolean().exclude('converted').sum())
.with_columns((pl.all() / c.processed).round(4).name.suffix('_pct'))
.select(cs.matches('.*_pct').exclude('processed_pct'))
.collect(engine='streaming')
.to_pandas()
)

compliant_pct  has_drugs_pct  has_pricing_pct
0         0.8747         0.7092           0.3336

### Overall Summary Statistics
- 2,992 files were processed of which 87% were compliant with CMS requirements. 71% of the files had drugs listed and 34% had both drugs and corresponding prices.

## State based Summary Statistics

In [5]:
(
base_query()
.group_by("state")
.agg(cs.boolean().exclude('converted').sum())
.sort("state")
.collect(engine='streaming')
.to_pandas()
)

state  processed  compliant  has_drugs  has_pricing
0     AK         26         17         14            4
1     AL         74         66         47           15
2     AR         74         69         53           33
3     AZ         74         68         40           29
4     CA         74         66         57           28
5     CO         74         63         59           28
6     CT         32         22         12            0
7     DE         14         14          9            2
8     FL         74         66         47           33
9     GA         74         69         57           18
10    HI         25         25         23           15
11    IA         74         53         44           18
12    ID         46         35         29           18
13    IL         74         65         54           12
14    IN         74         64         46           16
15    KS         74         65         60           27
16    KY         74         61         44           21
17    LA         70         56         39           23
18    MA         65         56         38           15
19    MD         61         60         50            9
20    ME         33         32         28           14
21    MI         74         70         46           13
22    MN         74         68         66           15
23    MO         74         63         52           29
24    MS         74         48         39           25
25    MT         53         41         39           11
26    NC         74         66         60           47
27    ND         39         33         31           12
28    NE         74         64         59           46
29    NH         30         28         26           13
30    NJ         74         64         48            9
31    NM         43         38         33           19
32    NV         60         54         46           29
33    NY         74         59         48           25
34    OH         74         70         55           26
35    OK         74         64         53           15
36    OR         44         38         37            9
37    PA         74         69         47           30
38    RI         15         12          9            2
39    SC         74         72         53           26
40    SD         60         57         53           28
41    TN         74         70         47           38
42    TX         74         67         56           38
43    UT         33         28         22           15
44    VA         74         64         54           33
45    VT         13         11         11            9
46    WA         74         58         55           19
47    WI         74         71         64            2
48    WV         62         60         48           33
49    WY         22         18         15            4

In [6]:
(
base_query()
.group_by("state")
.agg(cs.boolean().exclude('converted').sum())
.with_columns((cs.matches('(?i)comp|has') / c.processed).round(4).name.suffix('_pct'))
.select(cs.matches('state|.*_pct').exclude('processed_pct'))
.sort("state")
.collect(engine='streaming')
.to_pandas()
)

state  compliant_pct  has_drugs_pct  has_pricing_pct
0     AK         0.6538         0.5385           0.1538
1     AL         0.8919         0.6351           0.2027
2     AR         0.9324         0.7162           0.4459
3     AZ         0.9189         0.5405           0.3919
4     CA         0.8919         0.7703           0.3784
5     CO         0.8514         0.7973           0.3784
6     CT         0.6875         0.3750           0.0000
7     DE         1.0000         0.6429           0.1429
8     FL         0.8919         0.6351           0.4459
9     GA         0.9324         0.7703           0.2432
10    HI         1.0000         0.9200           0.6000
11    IA         0.7162         0.5946           0.2432
12    ID         0.7609         0.6304           0.3913
13    IL         0.8784         0.7297           0.1622
14    IN         0.8649         0.6216           0.2162
15    KS         0.8784         0.8108           0.3649
16    KY         0.8243         0.5946           0.2838
17    LA         0.8000         0.5571           0.3286
18    MA         0.8615         0.5846           0.2308
19    MD         0.9836         0.8197           0.1475
20    ME         0.9697         0.8485           0.4242
21    MI         0.9459         0.6216           0.1757
22    MN         0.9189         0.8919           0.2027
23    MO         0.8514         0.7027           0.3919
24    MS         0.6486         0.5270           0.3378
25    MT         0.7736         0.7358           0.2075
26    NC         0.8919         0.8108           0.6351
27    ND         0.8462         0.7949           0.3077
28    NE         0.8649         0.7973           0.6216
29    NH         0.9333         0.8667           0.4333
30    NJ         0.8649         0.6486           0.1216
31    NM         0.8837         0.7674           0.4419
32    NV         0.9000         0.7667           0.4833
33    NY         0.7973         0.6486           0.3378
34    OH         0.9459         0.7432           0.3514
35    OK         0.8649         0.7162           0.2027
36    OR         0.8636         0.8409           0.2045
37    PA         0.9324         0.6351           0.4054
38    RI         0.8000         0.6000           0.1333
39    SC         0.9730         0.7162           0.3514
40    SD         0.9500         0.8833           0.4667
41    TN         0.9459         0.6351           0.5135
42    TX         0.9054         0.7568           0.5135
43    UT         0.8485         0.6667           0.4545
44    VA         0.8649         0.7297           0.4459
45    VT         0.8462         0.8462           0.6923
46    WA         0.7838         0.7432           0.2568
47    WI         0.9595         0.8649           0.0270
48    WV         0.9677         0.7742           0.5323
49    WY         0.8182         0.6818           0.1818